django/utils/sys.py

```diff
@@ -0,0 +1,11 @@
+def reraise(exc_type, exc_value, tb=None):
+    """re raise excpetions captured by sys.exec_info()"""
+    try:
+        if exc_value is None:
+            exc_value = exc_type()
+        if exc_value.__traceback__ is not tb:
+            raise exc_value.with_traceback(tb)
+        raise exc_value
+    finally:
```

Comment:
I don't see the purpose of the try/finally (this is not in `six.reraise()` and no tests fail if it's removed. Further, the if statements are never evaluating to True.

Reply:
Looking up for six project, I ended on [bitbucket](https://bitbucket.org/gutworth/six/src/92e1c746e1abfbdb94705b821bc93766083efc54/six.py?at=default&fileviewer=file-view-default) former repository,
It explain the discrepancy, I will fix it asap.
Thank you.

Reply:
looks like upstream six repo is still using the same implementation.
I wouldn't change it. would you ?

Reply:
> no tests fail if it's removed.

it fails for me.
```
Traceback (most recent call last):
  File "tests/runtests.py", line 16, in <module>
    from django.test import TestCase, TransactionTestCase
  File "/home/nicolas/src/django/django/test/__init__.py", line 3, in <module>
    from django.test.client import Client, RequestFactory
  File "/home/nicolas/src/django/django/test/client.py", line 28, in <module>
    from django.utils.sys import reraise
ImportError: cannot import name 'reraise'
```
I will add a test for `reraise()` explicitly though.

Reply:
What I meant is that if you change the implementation to: 
```
def reraise(exc_type, exc_value, tb=None):
    raise exc_value
```
the new `test_client` test still passes. It also passes with `raise exc_info[1].with_traceback(exc_info[2])` as Claude suggested in the ticket. I think there should be additional assertions in the `test_client` test to demonstrate which implementation is correct.

Reply:
>  It also passes with raise exc_info[1].with_traceback(exc_info[2]) 

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.5/unittest/case.py", line 600, in run
    testMethod()
  File "/home/nicolas/src/django/tests/test_client/tests.py", line 730, in test_response_raise_exception
    self.client.get('/error/')
  File "/home/nicolas/src/django/django/test/client.py", line 518, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File "/home/nicolas/src/django/django/test/client.py", line 333, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/home/nicolas/src/django/django/test/client.py", line 405, in generic
    return self.request(**r)
  File "/home/nicolas/src/django/django/test/client.py", line 486, in request
    reraise(*exc_info)
  File "/home/nicolas/src/django/django/utils/sys.py", line 3, in reraise
    raise exc_type(exc_value).with_traceback(tb)
TypeError: __init__() missing 1 required positional argument: 'also_required'
```
Also with the test I backported from six (last commit), the `reraise` function is fully covered

Reply:
Yes, but the if statements don't evaluate to True in the test_client test so it's unclear whether they are  actually needed.

Reply:
Would you take the risk of using something else than `six` is doing ?
What if other corner cases comes up later for someone else.
This test proves that `raise exc_info[0](exc_info[1]).with_traceback(exc_info[2])` is not bullet proof. I would rather rely on a more trusted implementation from `six`, even if we don't know how to reproduce all the corner cases within the test_client test suite.
We know the actual implementation is fragile, since the fix is not trivial, I 'm incline to blindly trust `six.reraise` implementation since it is what was used before 2.0.

Reply:
I'd like to understand the code (and simplify, if possible) rather than include code that may not be necessary. I'll take a look if you don't want to spend more time on the issue.

Reply:
I don't want to make your work more difficult. please proceed.